### ISSUE TITLE
Ignore not serializable attributes

### ISSUE THREAD

**Issue by alexanderzuban** [k4tan/NHessian#3]
Propose to change ObjectSerializer to ignore read only or marked as "not serializable" fields.

```
 private static FieldInfo[] CollectFields(Type type)
        {
            var fields = new List<FieldInfo>();
            for (; type != null; type = type.BaseType)
            {
                fields.AddRange(
                    type.GetFields(
                        BindingFlags.Public |
                        BindingFlags.Instance |
                        BindingFlags.NonPublic |
                        BindingFlags.GetField |
                        BindingFlags.DeclaredOnly)
                    .Where(f => !f.IsNotSerialized && !f.IsLiteral && !f.IsInitOnly)
                    .ToList());
            }
            return fields.ToArray();
        }
```

**Comment by k4tan** [comment]
@alexanderzuban  Thanks for the proposal. Sounds like a good idea to me.

I think the `f.IsNotSerialized` bit seems pretty uncontroversal.

For the other two, I have two questions

1. Should literal/readonly be excluded for serialization, deserialization or both.
Could I serialize them but ignore on deserialization?

2. The deserialization should probably be considered as well `ObjectDeserializer.GetField`. 
I think that should get the same filter?

```csharp
// ObjectSerializer.cs
private static FieldInfo[] CollectFields(Type type)
{
    var fields = new List<FieldInfo>();
    for (; type != null; type = type.BaseType)
    {
        fields.AddRange(
            type.GetFields(
                BindingFlags.Public |
                BindingFlags.Instance |
                BindingFlags.NonPublic |
                BindingFlags.GetField |
                BindingFlags.DeclaredOnly)
            .Where(f => !f.IsNotSerialized)                // <<===
            .ToList()); 
    }
    return fields.ToArray();
}


// ObjectDeserializer.cs
private FieldInfo GetField(string fieldName)
{
    var type = MapType;
    for (; type != null; type = type.BaseType)
    {
        var fields = type.GetFields(
            BindingFlags.Public |
            BindingFlags.Instance |
            BindingFlags.NonPublic |
            BindingFlags.GetField |
            BindingFlags.DeclaredOnly)
            .Where(f => !f.IsNotSerialized && !f.IsLiteral && !f.IsInitOnly)   // <<===
            .ToList();

        for (int i = 0; i < fields.Count; i++)
            if (fields[i].Name == fieldName)
                return fields[i];
    }

    return null;
}
```

**Comment by alexanderzuban** [comment]
@k4tan 

It makes sense to have an implementation like yours if keep in mind the scenario to support legacy code. E.g. older model requires some fieldName and having "readonly fieldName  = ??;" will make it possible to support it. Anyway it is always possible to exclude a field with the  [NonSerialize] attribute.

**Comment by k4tan** [comment]
@alexanderzuban Alright. I am this week on holiday. Will take care of it once back.

**Comment by k4tan** [comment]
@alexanderzuban I found a bit of time and created #4 . 

The PR ended up a bit bigger as I wanted to make the behavior customizable by the user in case the default behavior is not suitable.

Feel free to have a look and give feedback.

I will merge once I have added some documentation to README.md.

**Comment by k4tan** [comment]
Released as part of [NHessian 0.3.0](https://github.com/k4tan/NHessian/releases/tag/v0.3)
